libdnf/dnf-swdb-pkg.c

```diff
@@ -0,0 +1,359 @@
+    swdbpkg->checksum_type = g_strdup (checksum_type);
+    swdbpkg->nevra = _pkg_make_nevra (swdbpkg);
+    swdbpkg->type = type;
+    return swdbpkg;
+}
+
+/**
+ * dnf_swdb_pkg_get_reason:
+ * @self: package object
+ *
+ * Returns: package reason ID
+ **/
+DnfSwdbReason
+dnf_swdb_pkg_get_reason (DnfSwdbPkg *self)
+{
+    DnfSwdbReason reason = DNF_SWDB_REASON_UNKNOWN;
+    if (!self || dnf_swdb_open (self->swdb)) {
+        return reason;
+    }
+
+    if (self->pid) {
+        reason = _reason_by_pid (self->swdb->db, self->pid);
+    }
+    return reason;
+}
+
+/**
+ * dnf_swdb_pkg_ui_from_repo:
+ * @self: package object
+ *
+ * Get UI reponame for package @self
+ * If package was installed with some other releasever, then
+ * print repository name in format @repo/releasever
+ *
+ * Returns: repository UI name
+ **/
+gchar *
+dnf_swdb_pkg_ui_from_repo (DnfSwdbPkg *self)
+{
+    if (self->_ui_repo)
+        return g_strdup (self->_ui_repo);
+    if (!self->swdb || !self->pid || dnf_swdb_open (self->swdb)) {
+        return g_strdup ("unknown");
+    }
+    sqlite3_stmt *res;
+    const gchar *sql = S_REPO_FROM_PID;
+    gint pdid = 0;
+    DB_PREP (self->swdb->db, sql, res);
+    DB_BIND_INT (res, "@pid", self->pid);
+    g_autofree gchar *r_name = NULL;
+    if (sqlite3_step (res) == SQLITE_ROW) {
+        r_name = g_strdup ((const gchar *)sqlite3_column_text (res, 0));
+        pdid = sqlite3_column_int (res, 1);
+    }
+    sqlite3_finalize (res);
+
+    // now we find out if package wasnt installed from some other releasever
+    if (pdid && self->swdb->releasever) {
+        g_autofree gchar *cur_releasever = NULL;
+        sql = S_RELEASEVER_FROM_PDID;
+        DB_PREP (self->swdb->db, sql, res);
+        DB_BIND_INT (res, "@pdid", pdid);
+        if (sqlite3_step (res) == SQLITE_ROW) {
+            cur_releasever = g_strdup ((const gchar *)sqlite3_column_text (res, 0));
+        }
+        sqlite3_finalize (res);
+        if (cur_releasever && g_strcmp0 (cur_releasever, self->swdb->releasever)) {
+            self->_ui_repo = g_strjoin (NULL, "@", r_name, "/", cur_releasever, NULL);
+            return g_strdup (self->_ui_repo);
+        }
+    }
+    if (r_name) {
+        self->_ui_repo = g_strjoin(NULL, "@", r_name, NULL);
+        return g_strdup (self->_ui_repo);
+    }
+    return g_strdup ("unknown");
+}
+
+/**
+ * dnf_swdb_pkg_compare:
+ * @pkg1: first package
+ * @pkg2: second package
+ *
+ * Performs "@pkg2 - @pkg1"
+ *
+ * Returns: < 0 when is @pkg1 newer, > 0 when older, else 0
+ **/
+gint64
+dnf_swdb_pkg_compare (DnfSwdbPkg *pkg1, DnfSwdbPkg *pkg2)
+{
+    // compare epochs
+    gint64 res = pkg2->epoch - pkg1->epoch;
+    if (res) {
+        return res;
+    }
+
+    // compare versions
+
+    // split version string into substrings
+    gchar **version1 = g_strsplit (pkg1->version, ".", 0);
+    gchar **version2 = g_strsplit (pkg2->version, ".", 0);
+
+    if (!version1 || !version2) {
+        return 0;
```

Comment:
Probably memory leak.
Eg. if version1!=NULL and version2==NULL we need to call g_strfreev(version1)

Reply:
Will fix